### PR TITLE
Unicode support in format strings

### PR DESCRIPTION
**Pull request by sirrahd**
This change fixes an issue where the Datetime Format GUI fails to load with the error 'Failed to convert locale string to UTF8: Invalid byte sequence in conversion input' if unicode characters are entered in any of the format strings.

Date().toLocaleFormat(format) only accepts ascii characters in the format string. This change works around that limitation by escaping non-ascii characters in the format string before passing the string to toLocaleFormat(), then escaping the result in case non-ascii characters are in the localized string, and finally unescaping the resulting string by passing it to JSON.parse.

### CODE REVIEW

File: src/Utilities.js [hunk 1]
```diff
@@ -12,7 +12,13 @@ const Gtk = imports.gi.Gtk;
 /// @return {string} Datetime representation of format, or format if the conversion fails, or datetime representation of defaultFormat, or blank.
 ///
 function dateTimeFormat(format, defaultFormat) {
-	return (format && new Date().toLocaleFormat(format) || format) || defaultFormat && new Date().toLocaleFormat(defaultFormat) || "";
+	function escapeUnicode(str) {
```
Comment (Daniel-Khodabakhsh) [inline src/Utilities.js]: Can you move this function out and add comments like the other functions in this file? And could you also rename the 'str' parameter to 'string'?

File: src/Utilities.js [hunk 2]
```diff
@@ -12,7 +12,13 @@ const Gtk = imports.gi.Gtk;
...
+		return str.replace(/[^\0-~]/g, function(ch) {
+			return "\\u" + ("0000" + ch.charCodeAt().toString(16)).slice(-4);
+		});
+	}
+
+	return JSON.parse('"' + escapeUnicode((format && new Date().toLocaleFormat(escapeUnicode(format)) || format) || defaultFormat && new Date().toLocaleFormat(escapeUnicode(defaultFormat)) || "").replace('"',     '\\"') + '"');
```
Comment (Daniel-Khodabakhsh) [inline src/Utilities.js]: Just wanted to test this before I merge it, want to make sure all edge cases are accounted for.

**Review by Daniel-Khodabakhsh — CHANGES_REQUESTED** [review]

**Comment by sirrahd** [comment]
Thanks Daniel. I did some edge case testing just now on this and found this doesn't handle " and \ characters well, need to escape them to make JSON.parse happy. I will push an update today with these fixes and your requested changes.

**Comment by sirrahd** [comment]
Just pushed an update that addresses your feedback and fixes the " and \ issues.
